File: frontend/src/types/results26.ts
/** Domain types for results / ResultRecord26 */
export type ResultRecord26Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface ResultRecord26 {
  id: string;
  externalCode: string;
  displayName: string;
  status: ResultRecord26Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface ResultRecord26CreateRequest {
  externalCode: string;
  displayName: string;
  status: ResultRecord26Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface ResultRecord26ListResponse {
  items: ResultRecord26[];
  total?: number;
}

export interface ResultRecord26StatsResponse {
  activeCount: number;
}

export function isResultRecord26Active(rec: ResultRecord26): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatResultRecord26Label(rec: ResultRecord26): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareResultRecord26ByPriority(a: ResultRecord26, b: ResultRecord26): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/results_26.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// ResultRecord26Service encapsulates business rules for ResultRecord26.
type ResultRecord26Service struct {
	repo *repository.ResultRecord26Repository
}

// NewResultRecord26Service wires the service.
func NewResultRecord26Service(repo *repository.ResultRecord26Repository) *ResultRecord26Service {
	return &ResultRecord26Service{repo: repo}
}

// Create validates and persists a new ResultRecord26.
func (s *ResultRecord26Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.ResultRecord26, error) {
	now := time.Now().UTC()
	rec := &models.ResultRecord26{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a ResultRecord26 by id.
func (s *ResultRecord26Service) Get(ctx context.Context, id string) (*models.ResultRecord26, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped ResultRecord26 pages.
func (s *ResultRecord26Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.ResultRecord26, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing ResultRecord26.
func (s *ResultRecord26Service) Update(ctx context.Context, actor string, rec *models.ResultRecord26) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a ResultRecord26.
func (s *ResultRecord26Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds ResultRecord26 by external code prefix.
func (s *ResultRecord26Service) Search(ctx context.Context, facilityID, prefix string) ([]models.ResultRecord26, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *ResultRecord26Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
